File: web/src/assvtt.test.ts
import { describe, expect, it } from 'vitest'
import { convertAssCue, parseAssHeader, positionDialogueCues } from './assvtt'

const HEADER = [
  '[Script Info]',
  'ScriptType: v4.00+',
  'PlayResX: 1920',
  'PlayResY: 1080',
  '',
  '[V4+ Styles]',
  'Format: Name, Fontname, Fontsize, PrimaryColour, SecondaryColour, OutlineColour, BackColour,'
    + ' Bold, Italic, Underline, StrikeOut, ScaleX, ScaleY, Spacing, Angle, BorderStyle,'
    + ' Outline, Shadow, Alignment, MarginL, MarginR, MarginV, Encoding',
  'Style: Default,Lato,48,&H00FFFFFF,&H000000FF,&H00000000,&H00000000,0,0,0,0,100,100,0,0,1,2,2,2,60,60,40,1',
  'Style: Signs,Lato,48,&H0000FFFF,&H000000FF,&H00000000,&H00000000,0,0,0,0,100,100,0,0,1,2,2,8,60,60,40,1',
  'Style: Thoughts,Lato,48,&H00FFFFFF,&H000000FF,&H00000000,&H00000000,0,-1,0,0,100,100,0,0,1,2,2,2,60,60,40,1',
].join('\n')

describe('convertAssCue', () => {
  const info = parseAssHeader(HEADER)

  it('places a top-anchored style at the top of the frame in its color', () => {
    expect(convertAssCue(info, 'Signs', 'placa')).toEqual({
      settings: 'line:5%',
      text: '<c.yellow>placa</c>',
    })
  })

  it('honors an inline alignment override', () => {
    expect(convertAssCue(info, 'Default', '{\\an8}no topo')).toEqual({
      settings: 'line:5%',
      text: 'no topo',
    })
  })

  it('reads a legacy alignment override the SSA way', () => {
    expect(convertAssCue(info, 'Default', '{\\a6}título').settings).toBe('line:5%')
  })

  it('turns \\pos into percentages of the script frame', () => {
    expect(convertAssCue(info, 'Default', '{\\pos(480,270)}placa solta').settings)
      .toBe('line:19% position:25%')
  })

  it('anchors a positioned top-aligned cue by its top edge', () => {
    expect(convertAssCue(info, 'Signs', '{\\pos(960,108)}alto').settings)
      .toBe('line:10% position:50%')
  })

  it('uses the first point of a movement', () => {
    expect(convertAssCue(info, 'Signs', '{\\move(960,108,0,0)}indo').settings)
      .toBe('line:10% position:50%')
  })

  it('aligns the side columns left and right', () => {
    expect(convertAssCue(info, 'Default', '{\\an7}canto').settings).toBe('line:5% position:5% align:left')
    expect(convertAssCue(info, 'Default', '{\\an3}outro').settings).toBe('position:95% align:right')
  })

  it('leaves bottom-center dialogue without settings', () => {
    expect(convertAssCue(info, 'Default', 'fala comum').settings).toBe('')
  })

  it('quantizes an inline color to the nearest VTT color class', () => {
    expect(convertAssCue(info, 'Default', '{\\c&H4020E0&}sangue').text).toBe('<c.red>sangue</c>')
  })

  it('closes an inline color when the override resets to the style', () => {
    expect(convertAssCue(info, 'Default', '{\\c&HFFFF00&}mar {\\c}céu').text)
      .toBe('<c.cyan>mar </c>céu')
  })

  it('returns every toggle to the style on \\r', () => {
    expect(convertAssCue(info, 'Thoughts', '{\\i0}dito {\\r}pensado').text)
      .toBe('dito <i>pensado</i>')
  })

  it('starts from the style for italics', () => {
    expect(convertAssCue(info, 'Thoughts', 'pensamento').text).toBe('<i>pensamento</i>')
  })

  it('drops vector drawings entirely', () => {
    expect(convertAssCue(info, 'Signs', '{\\p1}m 0 0 l 100 0 100 100{\\p0}').text).toBe('')
  })

  it('escapes markup-significant characters', () => {
    expect(convertAssCue(info, 'Default', 'a < b & b > c').text).toBe('a &lt; b &amp; b &gt; c')
  })

  it('handles hard breaks, soft breaks and hard spaces', () => {
    expect(convertAssCue(info, 'Default', 'um\\Ndois\\ntrês\\hquatro').text)
      .toBe('um\ndois três quatro')
  })

  it('falls back to plain bottom-center when the style is unknown', () => {
    expect(convertAssCue(info, 'Missing', 'fala')).toEqual({ settings: '', text: 'fala' })
  })

  it('reads styles from a legacy SSA header', () => {
    const legacy = [
      '[Script Info]',
      'ScriptType: v4.00',
      '[V4 Styles]',
      'Format: Name, Fontname, Fontsize, PrimaryColour, SecondaryColour, TertiaryColour,'
        + ' BackColour, Bold, Italic, BorderStyle, Outline, Shadow, Alignment,'
        + ' MarginL, MarginR, MarginV, AlphaLevel, Encoding',
      'Style: Top,Arial,20,65535,255,0,0,0,0,1,2,2,6,10,10,10,0,1',
    ].join('\n')
    expect(convertAssCue(parseAssHeader(legacy), 'Top', 'aviso')).toEqual({
      settings: 'line:5%',
      text: '<c.yellow>aviso</c>',
    })
  })
})

describe('positionDialogueCues', () => {
  it('lifts default dialogue off the bottom edge and leaves positioned cues alone', () => {
    const vtt = 'WEBVTT\n\n00:01.000 --> 00:02.000\nfala\n\n00:05.000 --> 00:06.000 line:5%\nplaca\n'

    const out = positionDialogueCues(vtt)

    expect(out).toContain('00:01.000 --> 00:02.000 line:-3\nfala')
    expect(out).toContain('00:05.000 --> 00:06.000 line:5%\nplaca')
  })

  it('sends the second simultaneous dialogue to the top of the frame', () => {
    const vtt = 'WEBVTT\n\n00:01.000 --> 00:04.000\nprimeira\n\n00:02.000 --> 00:05.000\nsegunda\n'

    const out = positionDialogueCues(vtt)

    expect(out).toContain('00:01.000 --> 00:04.000 line:-3\nprimeira')
    expect(out).toContain('00:02.000 --> 00:05.000 line:2\nsegunda')
  })

  it('returns to the bottom once the earlier dialogue has ended', () => {
    const vtt = 'WEBVTT\n\n00:01.000 --> 00:02.000\numa\n\n00:03.000 --> 00:04.000\noutra\n'

    const out = positionDialogueCues(vtt)

    expect(out).toContain('00:01.000 --> 00:02.000 line:-3\numa')
    expect(out).toContain('00:03.000 --> 00:04.000 line:-3\noutra')
  })

  it('reads the hour-bearing stamps our own converters write', () => {
    const out = positionDialogueCues('WEBVTT\n\n01:02:03.000 --> 01:02:04.000\ntarde\n')

    expect(out).toContain('01:02:03.000 --> 01:02:04.000 line:-3\ntarde')
  })
})

File: internal/media/assvtt_test.go
package media

import (
	"strings"
	"testing"

	"github.com/stretchr/testify/require"
)

func TestConvertASSToVTTKeepsPlacementAndColor(t *testing.T) {
	doc := strings.Join([]string{
		"[Script Info]",
		"ScriptType: v4.00+",
		"PlayResX: 1920",
		"PlayResY: 1080",
		"",
		"[V4+ Styles]",
		"Format: Name, Fontname, PrimaryColour, Bold, Italic, Alignment",
		"Style: Default,Lato,&H00FFFFFF,0,0,2",
		"Style: Signs,Lato,&H0000FFFF,0,0,8",
		"Style: Thoughts,Lato,&H00FFFFFF,0,-1,2",
		"",
		"[Events]",
		"Format: Layer, Start, End, Style, Name, MarginL, MarginR, MarginV, Effect, Text",
		"Dialogue: 0,0:00:01.00,0:00:02.50,Default,,0,0,0,,Uma fala, comum",
		"Dialogue: 0,0:00:03.00,0:00:04.00,Signs,,0,0,0,,{\\pos(960,108)}Ateliê",
		"Dialogue: 0,0:00:05.00,0:00:06.00,Thoughts,,0,0,0,,{\\i0}dito {\\r}pensado",
		"Dialogue: 0,0:00:07.00,0:00:08.00,Default,,0,0,0,,{\\an8}no topo\\Nsegunda",
		"Dialogue: 0,0:00:09.00,0:00:10.00,Default,,0,0,0,,{\\c&H4020E0&}sangue{\\c} e água",
		"Dialogue: 0,0:00:11.00,0:00:12.00,Signs,,0,0,0,,{\\p1}m 0 0 l 10 0{\\p0}",
		"Dialogue: 0,0:00:15.00,0:00:16.00,Default,,0,0,0,,{\\pos(480,270)}placa baixa",
		"Comment: 0,0:00:13.00,0:00:14.00,Default,,0,0,0,,ignorado",
	}, "\n")

	vtt := string(ConvertASSToVTT([]byte(doc)))

	require.True(t, strings.HasPrefix(vtt, "WEBVTT\n"))
	require.Contains(t, vtt, "00:00:01.000 --> 00:00:02.500 line:-3\nUma fala, comum\n")
	require.Contains(t, vtt, "00:00:03.000 --> 00:00:04.000 line:10% position:50%\n<c.yellow>Ateliê</c>\n")
	require.Contains(t, vtt, "00:00:05.000 --> 00:00:06.000 line:-3\ndito <i>pensado</i>\n")
	require.Contains(t, vtt, "00:00:07.000 --> 00:00:08.000 line:5%\nno topo\nsegunda\n")
	require.Contains(t, vtt, "00:00:09.000 --> 00:00:10.000 line:-3\n<c.red>sangue</c> e água\n")
	require.Contains(t, vtt, "00:00:15.000 --> 00:00:16.000 line:19% position:25%\nplaca baixa\n")
	require.NotContains(t, vtt, "m 0 0")
	require.NotContains(t, vtt, "ignorado")
}

func TestConvertASSToVTTEscapesMarkup(t *testing.T) {
	doc := strings.Join([]string{
		"[Events]",
		"Format: Layer, Start, End, Style, Name, MarginL, MarginR, MarginV, Effect, Text",
		"Dialogue: 0,0:00:01.00,0:00:02.00,Default,,0,0,0,,a < b & b > c",
	}, "\n")

	vtt := string(ConvertASSToVTT([]byte(doc)))

	require.Contains(t, vtt, "a &lt; b &amp; b &gt; c")
}

func TestConvertASSToVTTSortsCuesAndSkipsATruncatedTail(t *testing.T) {
	doc := strings.Join([]string{
		"[Events]",
		"Format: Layer, Start, End, Style, Name, MarginL, MarginR, MarginV, Effect, Text",
		"Dialogue: 0,0:00:05.00,0:00:06.00,Default,,0,0,0,,depois",
		"Dialogue: 0,0:00:01.00,0:00:02.00,Default,,0,0,0,,antes",
		"Dialogue: 0,0:00:07.00,0:0",
	}, "\n")

	vtt := string(ConvertASSToVTT([]byte(doc)))

	require.Less(t, strings.Index(vtt, "antes"), strings.Index(vtt, "depois"))
	require.NotContains(t, vtt, "00:00:07")
}

func TestConvertASSToVTTReturnsNothingWithoutCues(t *testing.T) {
	require.Empty(t, ConvertASSToVTT([]byte("[Script Info]\nTitle: vazio\n")))
}

func TestPositionDialogueCues(t *testing.T) {
	vtt := "WEBVTT\n\n" +
		"00:01.000 --> 00:04.000\nprimeira\n\n" +
		"00:00:02.000 --> 00:00:05.000\nsegunda\n\n" +
		"00:00:07.000 --> 00:00:08.000 line:5%\nplaca\n"

	out := string(positionDialogueCues([]byte(vtt)))

	require.Contains(t, out, "00:01.000 --> 00:04.000 line:-3\nprimeira")
	require.Contains(t, out, "00:00:02.000 --> 00:00:05.000 line:2\nsegunda")
	require.Contains(t, out, "00:00:07.000 --> 00:00:08.000 line:5%\nplaca")
}
